Improve inference of file mode when writing

mbhall88 opened the pull request:
The current method for inferring the file mode has a few issues where the correct mode gets overwritten.

https://github.com/brentp/cyvcf2/blob/d46fdcca8a3cb08b2ed4097199cb4632790c581f/cyvcf2/cyvcf2.pyx#L168-L173

in this situation, if the filename is `out.bcf.gz` then the first branch of the `if` statement is taken and the mode is `"wz"` which is compressed VCF. Additionally, if the filename is `out.bcf`, then the mode is `"wb"` which is *compressed* BCF - uncompressed BCF is `"wbu"` <sup>[1]</sup>. BUT, because of line 173, all of these settings are undone anyway, and the original `mode` passed into the function is used, which by default is `"w"`.   
**So in the end, regardless of the filename given, uncompressed VCF is written.**

---

This PR aims to improve the logic used for inferring the write mode from a filename. I have tested it locally for all 4 possible modes and according to [`htsfile`][2] it works as expected.

In addition:
-  I have updated the docstring for `Writer` to be explicit about what `mode` options are.
- Added a line to install requirements when building locally
- Clarified the FILTER "pass" possibilities

[1]: https://github.com/samtools/bcftools/blob/76392b3014de70b7fa5c6b5c9d5bc47361951770/version.c#L64-L70
[2]: https://www.htslib.org/doc/htsfile.html

mbhall88 commented:
> Thanks for addressing the problem. The function should also address e.g. wb7 or wz5 to indicate compression level. Could you verify that still works and update the docstring?

Good point, I didn't realise this. This also raises a concern I had with the current method. If someone provides a `mode` explicitly with a filename, then the mode will effectively be ignored. I see one of two options for addressing this:
1. Handle all of the file mode inference within `Writer` and then pass the "final" `mode` on to `_open_htsfile`. This would reduce the amount of work done in the `_open_htsfile` function, which makes sense as it seems more like the responsibility of the `Writer` to figure out what the mode should be.
2. A more "hacky" version why we keep the current model and if the mode is writing but is something other than `"w"` we blindly accept it and don't try and infer it from the filename?

Open to other ideas.

brentp commented:
I think we should only mess with mode if mode == "w" or mode == "r". If they specify anything beyond those, then they know what they are doing.

mbhall88 commented:
> I think we should only mess with mode if mode == "w" ~or mode == "r"~. If they specify anything beyond that, then they know what they are doing.

Ok. So to be clear, don't move mode inference to the `Writer`?

brentp commented:
> > I think we should only mess with mode if mode == "w" ~or mode == "r"~. If they specify anything beyond that, then they know what they are doing.
> 
> Ok. So to be clear, don't move mode inference to the `Writer`?

Let's move the inference to `Writer`. That's better. Thanks!

mbhall88 commented:
> Thanks for addressing the problem. The function should also address e.g. wb7 or wz5 to indicate compression level. Could you verify that still works and update the docstring?

Ok. So I have created files with varying modes (the mode is the filename) and checked what they are with `htsfile`

```
wb8.vcf:        BCF version 2.2 compressed variant calling data
wbu.vcf:        BCF version 2.2 variant calling data
wb.vcf: BCF version 2.2 compressed variant calling data
w.vcf:  VCF version 4.2 variant calling text
wz2.vcf:        VCF version 4.2 BGZF-compressed variant calling data
wz.vcf: VCF version 4.2 BGZF-compressed variant calling data
```

In addition, the file sizes indicate that the compressions level specification also works

```
446M  w.vcf
68M    wb.vcf
67M    wb8.vcf
356M  wbu.vcf
68M    wz.vcf
85M    wz2.vcf
```

I have also moved the mode inference into the `Writer` class and updated the docstring to hint at specifying compression level. I also added tests for the mode inference.

brentp commented:
For BCF, the default is compressed (they must specify mode including "u") to have uncompressed (unless they have .bcf.gz ext).
This matches htslib and common usage where most bcf files are compressed and have just a .bcf extension, not .bcf.gz

mbhall88 commented:
> For BCF, the default is compressed (they must specify mode including "u") to have uncompressed (unless they have .bcf.gz ext).
> This matches htslib and common usage where most bcf files are compressed and have just a .bcf extension, not .bcf.gz

Fair. Done. Also added a note about this in the docstring.

brentp commented:
thank you!
